RFC-0006-support-distributed-procedure.md

```diff
@@ -0,0 +1,431 @@
+
+```sql
+ALTER TABLE test_table EXECUTE optimize;
+ALTER TABLE test_table EXECUTE optimize(file_size_threshold => '128MB');
+ALTER TABLE test_partitioned_table EXECUTE optimize WHERE partition_key = 1;
+```
+
+Reference:
+* https://trino.io/docs/current/connector/iceberg.html#alter-table-execute
+
+## Goals
+
+* Expand the ability of procedures, enable them to leverage worker nodes to manipulate table data as well as metadata, which could be called as `distributed procedure`.
+* Expand currently procedure architecture to support the definition, registration, invocation and execution of `distributed procedures`.
+* Provide clarify architecture hierarchy and well defined SPI interface, enable various connectors to easily support distributed procedures.
+* Support distributed procedure for Iceberg connector, and implement a specific distributed procedure: `rewrite_data_files`.
+
+The ultimate goal of all of the above is that we can follow the original `CALL` syntax to execute procedures which have to be executed distributively, rather than extending SQL syntax.
+
+## Proposed Implementation
+
+### Overview
+
+1. Distributed procedure is used for processing of actual data besides metadata for a target table, such as data rewriting, merging, sorting or repartitioning. We need to re-factor and expand `Procedure/ProcedureRegistry` architecture, so that we can:
+   * Define a distributed procedure and register it into procedure registry
+   * Acquire and use procedure registry in `presto-analyzer` and `connectors` module
+
+> Since a distributed procedure is used to handle the data of a target table, it is enforced to contain two required parameters: `schema` and `table_name`.
+> Besides, it can also contain an optional parameter `filter` with the default value of `TRUE`, which indicates the filtering conditions for the data to be processed.
+> A specific procedure implementation can define the parameter list on its own, but it must include the required parameters `schema` and `table_name`, and can optionally use parameter `filter` to narrow down the range of data to be processed.
+
+2. Define a new query type `CALL_DISTRIBUTED_PROCEDURE`, and associate the call distributed procedure statement with this type during preparer phase. So that this kind of statements would fall into the path of `SqlQueryExecutionFactory.createQueryExecution(...)`. The generated `SqlQueryExecution` object can utilize the existing distributed querying, writing, and final committing mechanism to achieve distributed execution for procedures.
+
+
+3. Add a new plan node type: `CallDistributedProcedureNode`. During the analyzing and logical planning phase, construct a logical plan with the following shape for `call distributed procedure` statement:
+
+    ```text
+    TableScanNode -> FilterNode -> CallDistributedProcedureNode -> TableFinishNode -> OutputNode
```

Comment:
The operation may not necessarily update the real data files. For example, We may want to run `CALL system.clean_data_cache('ALL_WORKERS');` to just call `AsyncDataCache::shutdown() ` on all Presto C++ workers. What would the plan look like?

Reply:
cc @agrawalreetika